design/crd-status.md

```diff
@@ -0,0 +1,237 @@
+# Exposing the status with CRDs
+
+## Summary
+
+The purpose of this enhancement is to expose useful information to perform troubleshooting
+via CRDs.
+
+## Motivation
+
+Exposing part of the internal state was not possible until the introduction of CRD
+based configuration, and despite part of the status is exposed via Prometheus
+metrics, troubleshooting MetalLBoften requires inspecting the logs of the various
+controllers, and it's not always easy to understand why a service is not working.
+
+For example, inspecting logs is required when:
+
+* an invalid configuration was applied
+* the session with a BGP peer is not established
+* a service was not announced to a BGP peer
+
+### Goals
+
+* Expose useful information to ease the debug
+
+A high level list of informations we want to retrieve is:
+
+* configuration status (valid/invalid and reason)
+* number of used / available IPs per IPAddressPool
+* session state for both BGP and BFD, per node - peer
+* service announcement status, per peer - node (BGP)
+* service announcement node (L2)
+
+### Non-Goals
+
+* Providing BGP status for the legacy implementation
+* Providing a Status subsection for each resource
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to see from which nodes my service is
+announced, and to which `BGPpeer`s.
+
+#### Story 2
+
+As a cluster administrator, I want to know if the BGP / BFD session with a given
+peer is established or not, for each node.
+
+#### Story 3
+
+As a cluster administrator, I want to know if the configuration applied
+is valid or not, and why it failed.
+
+## Design Details
+
+### Making the solution scalable
+
+The biggest challenge is the fact that all the concepts related to MetalLB
+are cluster scoped, but a lot of the information we care about is node scoped.
+
+A clear example is the state of the BGP session established with a given
+`BGPpeer`, where the `BGPpeer` is defined as a cluster concept, but sessions
+are established from different nodes.
+
+If we would add a `Status` field to the `BGPPeer` resource, it would add an
+unwanted load to the APIServer on clusters with an high number of nodes,
+especially when dealing with faulty networks where the connectivity is intermittent.
```

Comment:
Can you please clarify how having a BGPSessionStatus CRD for each BGP peer (which can be thousands and more) 
is less overhead than adding this status to the already existing BGPPeer?

Reply:
First, thousands of bgppeers is not realistic (as opposed to thousands of nodes).
And the difference here is the size of each resource and the number of updates.

This is both about the size of the payload each session update will produce, the conflicts and the number of upgrades.
Think about 1000 nodes and two flickering sessions from two nodes (to the same peer).
Every flick will require to move a CR that is way bigger, the updates will be the combination of the two sessions (down down, down up, up up), not considerning the number of retries given by the fact that they may try to update their version at the same time. And this is with two sessions only.

Having one CR per session limits this, while not affecting the UX (too much).